pkgs/jnigen/docs/java_differences.md

```diff
@@ -0,0 +1,166 @@
+## Syntax and semantic differences between Java and the generated Dart bindings
+
+### Method overloading
+
+Java supports overloading methods. This means that it can distinguish between
+two methods with the same name that have a different signature.
+
+```java
+// Java
+public class Calculator {
+  public int add(int a, int b) {
+    return a + b;
+  }
+
+  public double add(double a, double b) {
+    return a + b;
+  }
+
+  public float add(float a, float b) {
+    return a + b;
+  }
+}
+```
+
+This is not the case for Dart. Each method of a class must have a unique name.
+To overcome this limitation, JNIgen adds a dollar sign (`$`) and a numeric
+suffix to the end of the overloaded method name.
+
+```dart
+// Dart Bindings - Boilerplate omitted for clarity.
+class Calculator extends JObject {
+  int add(int a, int b) { /* ... */ }
+
+  double add$1(double a, double b) { /* ... */ }
+
+  double add$2(double a, double b) { /* ... */ }
+}
+```
+
+> [!WARNING]  
+> Running the code generator again on a different version of the library can map
+> the methods differently if the order of the methods change or another
+> overloading of the same method gets added.
+>
+> Double check the doc comments on the generated methods to make sure you are
+> calling your intended method.
+
+You might wonder why the method isn't renamed to `add1` instead of `add$1`. The
+reason is that a method named `add1` could already exist in the Java class. On
+the other hand, Java identifiers cannot contain dollar signs.
```

Comment:
Java identifiers *can* contain `$`: https://docs.oracle.com/javase/specs/jls/se11/html/jls-3.html#jls-3.8

They *shouldn't*, and the character is morally reserved for code generation and referencing legacy code that blatantly used `$`.
However, Dart code may be used to interact with code-generated Java code, so it cannot assume that there are no `$`s in the source names.

I recommend either changing existing `$`s to `$$`, or always adding a suffix to names containing `$`, starting with `$0` where you would otherwise hav done nothing. Then any Dart name that ends with `$<digits>`, (or if using the `$`-doubling, one that is not preceded by an odd number of `$`s) is a JNIGen rename, and cannot conflict with an un-renamed name.

Reply:
🤦‍♂️ I didn't know this! Will add the logic, thanks